chore(docs): an additional doc on fold (#750)
fluencelabs/aquavm: docs/fold.md

## AIR: Fold instruction
Here is the `fold` syntax.

`(fold <iterable> <iterator> <instruction> [<last_instruction>])`

Iterable can be: a stream, map, canonicalized stream, canonicalized map or scalar array.

If `last_instruction` is not set, then it's assumed:
 - `null` in case of `(fold (seq ...`
 - `never` in case of `(fold (par ...`

Here is a simple example when `fold` loops over a scalar to populate map using `iterator` as key-value pair value:
```clojure
(fold iterable iterator
	(seq
	  (ap ("key" iterator) %map)
	  (next iterator)
	)
	;; here is an implicit (null)
)
```

> This example demonstrates the case when `next` is not the last instruction. This feature is supported with `fold`` over scalar only.

Reviewer: ```suggestion
This example demonstrates the case when `next` is not the last instruction. This feature is supported with `fold` over scalar only.
```

Author: Fixed

```clojure
(fold iterable iterator
	(seq
	  (next iterator)
	  (ap ("key" iterator) %map)
	)
	;; here is an implicit (null)
)
```

> Fold body can have no `next`. The body will be executed only once in this case.

Reviewer: I think it is important to emphasize that this is true for `fold` over scalars and canons only

```clojure
(fold iterable iterator
	(seq
	  (ap ("key" iterator) %map)
	  	(seq ;; next
		  (ap ("key" iterator) %map)
			(seq ;; next
				(ap ("key" iterator) %map)
				;; an implicit (null)
			)
		)
	)
)
```

There is so-called recursive fold which populates iterable stream `fold` is iterating over.

```clojure
(fold $iterable iterator ; data is a stream in the context
	(seq
	  (ap 42 $iterable)
	  (next iterator)
	)
	;; here is an implicit (null)
)
```

```clojure
(fold $iterable iterator ; data is a stream in the context
	(seq
	  (ap 42 $iterable)
		(seq ;; next
		  (ap 42 $iterable)
			(seq ;; next
				(ap 42 $iterable)
					...
			)
		)
	)
	;; here is an implicit (null)
)
```

A stream internally is a vector of vectors where inner vectors are called generations. In the previous case ap adds a new generation into `fold`.

Here is a high level overview how `fold` handles multiple generations of an iterable.

```clojure
fold
	run fold body for generation 1 values
		run last instruction
	run fold body for generation 2 values
		run last instruction
	...
	run fold body for generation N values
		run last instruction
```

> There are some subtle details about fold processing over a number of generations, namely:

Reviewer: ```suggestion
There are some subtle details about `fold` processing over a number of generations, namely:
```

- runs of previous generations do not affect next generations in any way

> - if previous generation doesn't call `next` AquaVM executes `fold` for the next generation.

Reviewer: ```suggestion
- even if previous generation doesn't call `next` AquaVM executes `fold` for the next generation.
```

## AIR: Fold instruction
Here is the `fold` syntax.

`(fold <iterable> <iterator> <instruction> [<last_instruction>])`

Iterable can be:
- scalars: canonicalized stream, canonicalized map, scalar array
- stream-based structs: stream, map

Please note that `fold`'s algo for a scalar iterable is a subset of the `algo` for stream-based iterable. See [`fold` for stream specifics](#fold-for-stream-specifics)

If `last_instruction` is not set, then it's assumed:
 - `null` in case of `(fold (seq ...`
 - `never` in case of `(fold (par ...`

Here is a simple example when `fold` loops over a scalar to populate map using `iterator` as key-value pair value:
```clojure
(fold iterable iterator
	(seq
	  (ap ("key" iterator) %map)
	  (next iterator)
	)
	;; here is an implicit (null)
)
```

This example demonstrates the case when `next` is not the last instruction. This feature is supported with `fold` over scalar only.
```clojure
(fold iterable iterator
	(seq
	  (next iterator)
	  (ap ("key" iterator) %map)
	)
	;; here is an implicit (null)
)
```

> Fold body can have no `next`. The body will be executed only once in this case.

Reviewer: ```suggestion
The `fold` body can have no `next`. The body will be executed only once in this case.
```

```clojure
(fold iterable iterator
	  (ap ("key" iterator) %map)
	;; here is an implicit (null)
)
```

> Let's consider the fold's operational semantics if iterable is the following scalar array `[1 2 3]`. Note that the recursion produced by `next` expressed in recursion of AquaVM execution engine function calls. Putting this differently the depth of the recursion depends on the size of the iterable.

Reviewer: ```suggestion
Let's consider the `fold`'s operational semantics if iterable is the following scalar array `[1 2 3]`. Note that the recursion produced by `next` expressed in recursion of AquaVM execution engine function calls. Putting this differently the depth of the recursion depends on the size of the iterable.
```

```clojure
(fold iterable iterator
	(seq
	  (ap ("key" iterator) %map)
	  	(seq ;; next
		  (ap ("key" iterator) %map)
			(seq ;; next
				(ap ("key" iterator) %map)
				;; an implicit (null)
			)
		)
	)
)
```

> There is so-called recursive fold which populates iterable stream `fold` is iterating over.

Reviewer: Why is this example not in the next "`fold` for stream" section?

```clojure
(fold $iterable iterator ; data is a stream in the context
	(seq
	  (ap 42 $iterable)
	  (next iterator)
	)
	;; here is an implicit (null)
)
```

```clojure
(fold $iterable iterator ; data is a stream in the context
	(seq
	  (ap 42 $iterable)
		(seq ;; next
		  (ap 42 $iterable)
			(seq ;; next
				(ap 42 $iterable)
					...
			)
		)
	)
	;; here is an implicit (null)
)
```

# `fold` for stream specifics
A stream internally is a vector of vectors where inner vectors are called generations. In the previous case ap adds a new generation into `fold`.

Here is a high level overview how `fold` handles multiple generations of an iterable.

```clojure
fold
	run fold body for generation 1 values
		run last instruction
	run fold body for generation 2 values
		run last instruction
	...
	run fold body for generation N values
		run last instruction
```
There are some subtle details about `fold` processing over a number of generations, namely:
- runs of previous generations do not affect next generations in any way

> - even if previous generation doesn't call `next` AquaVM executes `fold` for the next generation.

Reviewer: But if `next` is not called in one generation, other elements in this generation are not processed?

> There is a property of `fold` completeness. The invariant for this property is as follows, if fold finishes a run for at least one generation the fold is marked as complete. This property belongs to AquaVM execution engine and it is not explicitly sent out in a particle.

Reviewer: ```suggestion
There is a property of `fold` completeness. The invariant for this property is as follows, if fold finishes a run for at least one generation the `fold` is marked as complete. This property belongs to AquaVM execution engine and it is not explicitly sent out in a particle.
```

> Please note that fold body that triggers a node change events, e.g. `call`,`canon` destined for a node other than the current, forces the executiont to migrate to the destined node.

Reviewer: ```suggestion
Please note that `fold` body that triggers a node change events, e.g. `call`, `canon` destined for a node other than the current, forces the execution to migrate to the destined node.
```

## AIR: Fold instruction
Here is the `fold` syntax.

`(fold <iterable> <iterator> <instruction> [<last_instruction>])`

> 

Reviewer 2: Let add here that `Iterable` can be a stream, map, canonicalized stream, canonicalized map or scalar array.

Author: Fixed

`last_instruction` is:
- an implicit `null` for scalar-based iterables

> - an implicit `never` for stream-based iterables

Author: Fixed

```clojure
(fold data value
	(seq
	  (ap ("somesome-lengthy-and-scary-key" value) %map)
	  (next value)
	)
	;; here is an implicit (null)
)
```

> Here is an operational semantics of fold is if data is `[1 2 3]`. Please note that the recursion produced by `next` expressed in recursion of AquaVM execution engine function calls. Putting this differently the depth of the recursion depends on the size of the iterable.

Author: Fixed

```clojure
(fold data value
	(seq
	  (ap ("somesome-lengthy-and-scary-key" value) %map)
	  	(seq ;; next
		  (ap ("somesome-lengthy-and-scary-key" value) %map)
			(seq ;; next
				(ap ("somesome-lengthy-and-scary-key" value) %map)
				;; an implicit (null)
			)
		)
	)
)
```

There is so-called recursive fold which populates iterable stream `fold` is iterating over.

```clojure
(fold $data value ; data is a stream in the context
	(seq
	  (ap 42 $data)
	  (next value)
	)
	;; here is an implicit (null)
)
```

```clojure
(fold $data value ; data is a stream in the context
	(seq
	  (ap 42 $data)
		(seq ;; next
		  (ap 42 $data)
			(seq ;; next
				(ap 42 $data)
					...
			)
		)
	)
	;; here is an implicit (null)
)
```

> Fold internally is a vector of vectors where inner vectors are called generations. In the previous case ap adds a new generation into `fold`.

Reviewer 2: ```suggestion
A stream internally is a vector of vectors where inner vectors are called generations. In the previous case ap adds a new generation into `fold`.
```

Author: Fixed

## AIR: Fold instruction
Here is the `fold` syntax.

> `(fold <iterable> <iterator> <instruction> [<last_instruction>])`

Reviewer: How does `fold` handle errors in its `instruction`?

Reviewer 2: Only fold over stream influences error, actually it supresses them, otherwise it'd be not deterministic. Fold over scalar and canonicalized maps and streams doesn't influence errors anyhow, the best analogy is the provided one with a tree, so it behaves exactly like in this tree representation.

Reviewer 3: can we dump answers into documentation too?

Iterable can be: a stream, map, canonicalized stream, canonicalized map or scalar array.

If `last_instruction` is not set, then it's assumed:
 - `null` in case of `(fold (seq ...`
 - `never` in case of `(fold (par ...`

```clojure
(fold iterable iterator
	(seq
	  (ap ("key" iterator) %map)

> 	  (next iterator)

Reviewer: Would be nice to provide an example with `next` not as last instruction in `isntruction` also

Reviewer: And elaborate on what happens if `next` is **not** called inside `instruction`

Author: Good point

```clojure
(fold iterable iterator
	(seq
	  (ap ("key" iterator) %map)
	  	(seq ;; next
		  (ap ("key" iterator) %map)
			(seq ;; next
				(ap ("key" iterator) %map)
				;; an implicit (null)
			)
		)
	)
)
```

There is so-called recursive fold which populates iterable stream `fold` is iterating over.

```clojure
(fold $iterable iterator ; data is a stream in the context
	(seq
	  (ap 42 $iterable)
	  (next iterator)
	)
	;; here is an implicit (null)
)
```

```clojure
(fold $iterable iterator ; data is a stream in the context
	(seq
	  (ap 42 $iterable)
		(seq ;; next
		  (ap 42 $iterable)
			(seq ;; next
				(ap 42 $iterable)
					...
			)
		)
	)
	;; here is an implicit (null)
)
```

A stream internally is a vector of vectors where inner vectors are called generations. In the previous case ap adds a new generation into `fold`.

Here is a high level overview how `fold` handles multiple generations of an iterable.

```clojure
fold
	run fold body for generation 1 values
		run last instruction

> 	run fold body for generation 2 values

Reviewer: Runs of previous generations do not affect next generations in any way? 
What if previous generation didn't call `next`? Next generation will still be executed?

Reviewer 2: > Runs of previous generations do not affect next generations in any way?
It doesn't

What if previous generation didn't call next? Next generation will still be executed?
Yep, it doesn't depend on whether next was called or not.

		run last instruction
	...
	run fold body for generation N values
		run last instruction
```

> There is a property of `fold` completeness. The invariant for this property is as follows, if fold finishes a run for at least one generation the fold is marked as complete. This property belongs to AquaVM execution engine and it is not explicitly sent out in a particle.

Reviewer: But AquaVM tries to run other generations still?

Author: It does.
